src/core/payments/funding.ts

```diff
@@ -0,0 +1,426 @@
+        ? `Required funding for ${plan.targetRunwayDays} days of storage`
+        : 'Required funding to meet runway target'
+    case 'runway-with-piece':
+      return plan?.targetRunwayDays != null
+        ? `Required funding for ${plan.targetRunwayDays} days of storage (including upcoming upload)`
+        : 'Required funding for storage runway (including upcoming upload)'
+    case 'target-deposit':
+      return 'Required funding to reach target deposit amount'
+    case 'withdrawal-excess':
+      return 'Excess funds available for withdrawal'
+    default:
+      return 'Required funding'
+  }
+}
+
+/**
+ * Calculate a Filecoin Pay funding plan without making network calls
+ *
+ * This is a pure calculation function that determines what funding adjustments
+ * are needed to reach a target. Use this when you already have PaymentStatus
+ * and pricing information, or when you need synchronous calculation logic.
+ *
+ * For full workflow including network calls, allowance checks, and execution,
+ * use planFilecoinPayFunding instead.
+ *
+ * @param options - Calculation options with payment status
+ * @returns Funding plan with delta, action, and insights
+ */
+export function calculateFilecoinPayFundingPlan(options: FilecoinPayFundingPlanOptions): FilecoinPayFundingPlan {
+  const {
+    status,
+    targetRunwayDays,
+    targetDeposit,
+    pieceSizeBytes,
+    pricePerTiBPerEpoch,
+    mode = 'exact',
+    allowWithdraw = true,
+  } = options
+
+  if (targetRunwayDays != null && targetDeposit != null) {
+    throw new Error('Specify either targetRunwayDays or targetDeposit, not both')
+  }
+
+  if (targetRunwayDays == null && targetDeposit == null) {
+    throw new Error('A funding target is required')
+  }
+
+  if (pieceSizeBytes != null && pricePerTiBPerEpoch == null) {
+    throw new Error('pricePerTiBPerEpoch is required when pieceSizeBytes is provided')
+  }
+
+  let delta: bigint
+  let projectedDeposit = status.filecoinPayBalance
+  let projectedRateUsed = status.currentAllowances.rateUsed ?? 0n
+  let projectedLockupUsed: bigint
+  let resolvedTargetDeposit: bigint | undefined
+  let reasonCode: FundingReasonCode = 'none'
+  const targetType = targetRunwayDays != null ? 'runway-days' : 'deposit'
+
+  if (targetType === 'runway-days') {
+    const days = targetRunwayDays ?? 0
+    if (pieceSizeBytes != null) {
+      if (pricePerTiBPerEpoch == null) {
+        throw new Error('pricePerTiBPerEpoch is required when planning with pieceSizeBytes')
+      }
+      const adjustment = computeAdjustmentForExactDaysWithPiece(status, days, pieceSizeBytes, pricePerTiBPerEpoch)
+      delta = adjustment.delta
+      resolvedTargetDeposit = adjustment.targetDeposit
+      projectedRateUsed = adjustment.newRateUsed
+      projectedLockupUsed = adjustment.newLockupUsed
+
+      // Determine reason: piece upload with or without runway
+      if (days === 0) {
+        reasonCode = delta > 0n ? 'piece-upload' : 'none'
```

Comment:
what is this case? you want zero days runway but still need funding to onboard the piece? is this a normal path we experience?